### PR TITLE
Fix #477: output_transform parameter added to create_supervised... functions

### LINKED ISSUE #477: Improve create_supervised_trainer with optional output_transform
Following [the discussion](https://github.com/pytorch/ignite/pull/476#discussion_r272108999), idea is to give more flexibility to users who are using `create_supervised_trainer`:
```python
def default_output_transform(x, y, y_pred, loss):
    return loss.item() 


def create_supervised_trainer(model, optimizer, loss_fn,
                              device=None, non_blocking=False, prepare_batch=_prepare_batch, 
                              output_transform=default_output_transform):
    if device:
        model.to(device)

    def _update(engine, batch):
        model.train()
        optimizer.zero_grad()
        x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
        y_pred = model(x)
        loss = loss_fn(y_pred, y)
        loss.backward()
        optimizer.step()
        return output_transform(x, y, y_pred, loss)

    return Engine(_update)    
```

cc @IlyaOvodov

### PR DESCRIPTION
**Pull request by IlyaOvodov**
'output_transform' parameter added to create_supervised_trainer(...) and create_supervised_evaluator(...) functions.

It can be used to change value being assigned to engine.state.output
For instance, it can be changed from loss.item() to (y, pred_y) if one need to attach metrics to training process.

Fixes #477

### CODE REVIEW

File: ignite/engine/__init__.py [hunk 1]
```diff
@@ -29,6 +30,10 @@ def create_supervised_trainer(model, optimizer, loss_fn,
             with respect to the host. For other cases, this argument has no effect.
         prepare_batch (callable, optional): function that receives `batch`, `device`, `non_blocking` and outputs
             tuple of tensors `(batch_x, batch_y)`.
+        output_transform (callable, optional): function that receives 'x', 'y', 'y_pred', 'loss' and returns value
+            to be assigned to engine's state.output after each iteration. Default is returning loss.item(). If you
```
Comment (vfdev-5) [inline ignite/engine/__init__.py]:
```suggestion
            to be assigned to engine's state.output after each iteration. Default is returning `loss.item()`. If you
```

File: ignite/engine/__init__.py [hunk 2]
```diff
@@ -29,6 +30,10 @@ def create_supervised_trainer(model, optimizer, loss_fn,
...
+            are going to attach metrics to trainer, you should pass
```
Comment (vfdev-5) [inline ignite/engine/__init__.py]: nit: I would move "If you are going to attach metrics to trainer, you should pass `output_transform = lambda x, y, y_pred, loss: (y_pred, y,)`" to `create_supervised_evaluator`. We are not intended to attach metrics to the trainer as metrics as somewhat running averages of running model.

File: ignite/engine/__init__.py [hunk 3]
```diff
@@ -66,6 +72,8 @@ def create_supervised_evaluator(model, metrics={},
             with respect to the host. For other cases, this argument has no effect.
         prepare_batch (callable, optional): function that receives `batch`, `device`, `non_blocking` and outputs
             tuple of tensors `(batch_x, batch_y)`.
+        output_transform (callable, optional): function that receives 'x', 'y', 'y_pred' and returns value
+            to be assigned to engine's state.output after each iteration. Default is returning '(y_pred, y,)'
```
Comment (vfdev-5) [inline ignite/engine/__init__.py]:
```suggestion
            to be assigned to engine's state.output after each iteration. Default is returning `(y_pred, y,)`.
```

File: ignite/engine/__init__.py [hunk 4]
```diff
@@ -66,6 +72,8 @@ def create_supervised_evaluator(model, metrics={},
...
 
     Note: `engine.state.output` for this engine is a tuple of `(batch_pred, batch_y)`.
```
Comment (vfdev-5) [inline ignite/engine/__init__.py]: Probably we need to remove this line

File: ignite/engine/__init__.py (hunk 2 above)
Comment (IlyaOvodov) [reply ignite/engine/__init__.py]: No need to move it to create_supervised_evaluator because it's already OK there by default.
Changing output to (y_pred, y,) is required here and main motivation for it is attaching metrics to trainer. If loss calculation is time consuming, calculating metrics on train set while trainig is better choice then calculating them using evaluator because it means repeated run on the train set. Although the second option is more precise of cause, it's cost can be too high.

File: ignite/engine/__init__.py (hunk 4 above)
Comment (IlyaOvodov) [reply ignite/engine/__init__.py]: Do you mean  line
Note: `engine.state.output` for this engine is a tuple of `(batch_pred, batch_y)`.
Yes. But better to update it.

File: ignite/engine/__init__.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/engine/__init__.py]: @IlyaOvodov we do not encourage attaching metrics to the trainer : https://pytorch.org/ignite/master/quickstart.html#id1 
So, please remove this phrase. 
And in case of `create_supervised_evaluator` we can just recall that to compute metrics we need to return also `y_pred` and `y`.

File: ignite/engine/__init__.py (hunk 4 above)
Comment (vfdev-5) [reply ignite/engine/__init__.py]: Yes, please update it

File: ignite/engine/__init__.py (hunk 2 above)
Comment (IlyaOvodov) [reply ignite/engine/__init__.py]: Done. Also added comment on 'create_supervised_evaluator'.

File: ignite/engine/__init__.py [hunk 5]
```diff
@@ -66,8 +71,12 @@ def create_supervised_evaluator(model, metrics={},
             with respect to the host. For other cases, this argument has no effect.
         prepare_batch (callable, optional): function that receives `batch`, `device`, `non_blocking` and outputs
             tuple of tensors `(batch_x, batch_y)`.
+        output_transform (callable, optional): function that receives 'x', 'y', 'y_pred' and returns value
+            to be assigned to engine's state.output after each iteration. Default is returning `(y_pred, y,)` with fits
```
Comment (vfdev-5) [inline ignite/engine/__init__.py]:
```suggestion
            to be assigned to engine's state.output after each iteration. Default is returning `(y_pred, y,)` which fits
```

File: ignite/engine/__init__.py [hunk 6]
```diff
@@ -66,8 +71,12 @@ def create_supervised_evaluator(model, metrics={},
...
+            output expected by metrics. If you change it you should use `output_transform` in metrcis.
```
Comment (vfdev-5) [inline ignite/engine/__init__.py]:
```suggestion
            output expected by metrics. If you change it you should use `output_transform` in metrics.
```

File: ignite/engine/__init__.py (hunk 2 above)
Comment (IlyaOvodov) [reply ignite/engine/__init__.py]: By the way. What if to change default output_transform for metrics from lambda x: x to lambda x: x[:2]?
As metrics now experct tuple (y_pred, y) it will not change current functioning but extend it: tuple of (y_pred, y, somth else,...) will become acceptable, so adding some information to state.output using means of this PR will not require using output_transform in metrics.

Comment (vfdev-5) [reply ignite/engine/__init__.py]: Good idea, maybe something like `lambda x: x[:2] if len(x) > 2 else x` just to ensure

Comment (vfdev-5) [reply ignite/engine/__init__.py]: I think in v0.3.0 we should simplify this stuff and write a dict like `{'y_pred': y_pred, 'y': y}` to the `state.output`

Comment (IlyaOvodov) [reply ignite/engine/__init__.py]: :) 
Need to think about it. From one point of view it's good (I started with it), but somebody can use some custom metric inherited fromMetric class and expect to receive that output that he returned from process function, proposed sofistication will be suurprise...
